cpp/src/text/subword/data_normalizer.cu

```diff
@@ -0,0 +1,459 @@
+                                       uint32_t const* device_strings_offsets,
+                                       size_t const total_bytes,
+                                       uint32_t const* cp_metadata,
+                                       uint64_t const* aux_table,
+                                       bool const do_lower_case,
+                                       uint32_t const num_strings,
+                                       uint32_t* code_points,
+                                       uint32_t* chars_per_thread)
+{
+  constexpr uint32_t init_val                     = (1 << SORT_BIT);
+  uint32_t replacement_code_points[MAX_NEW_CHARS] = {init_val, init_val, init_val};
+
+  const uint32_t char_for_thread = blockDim.x * blockIdx.x + threadIdx.x;
+  uint32_t num_new_chars         = 0;
+
+  if (char_for_thread < total_bytes) {
+    auto const code_point = extract_code_points_from_utf8(strings, char_for_thread);
+    auto const metadata   = cp_metadata[code_point];
+
+    if (is_head_byte(strings[char_for_thread]) && !should_remove_cp(metadata, do_lower_case)) {
+      num_new_chars = 1;
+      // Apply lower cases and accent stripping if necessary
+      auto const new_cp =
+        do_lower_case || always_replace(metadata) ? get_first_cp(metadata) : code_point;
+      replacement_code_points[0] = new_cp == 0 ? code_point : new_cp;
+
+      if (do_lower_case && is_multi_char_transform(metadata)) {
+        auto const next_cps          = aux_table[code_point];
+        replacement_code_points[1]   = static_cast<uint32_t>(next_cps >> 32);
+        auto const potential_next_cp = static_cast<uint32_t>(next_cps);
+        replacement_code_points[2] =
+          potential_next_cp != 0 ? potential_next_cp : replacement_code_points[2];
+        num_new_chars = 2 + (potential_next_cp != 0);
+      }
+
+      if (should_add_spaces(metadata, do_lower_case)) {
+        // Need to shift all existing code-points up one
+        // This is a rotate right. There is no thrust equivalent at this time.
+        for (int loc = num_new_chars; loc > 0; --loc) {
+          replacement_code_points[loc] = replacement_code_points[loc - 1];
+        }
+
+        // Write the required spaces at the end
+        replacement_code_points[0]                 = SPACE_CODE_POINT;
+        replacement_code_points[num_new_chars + 1] = SPACE_CODE_POINT;
+        num_new_chars += 2;
+      }
+    }
+  }
+
+  chars_per_thread[char_for_thread] = num_new_chars;
+
+  typedef cub::
+    BlockStore<uint32_t, THREADS_PER_BLOCK, MAX_NEW_CHARS, cub::BLOCK_STORE_WARP_TRANSPOSE>
+      BlockStore;
+  __shared__ typename BlockStore::TempStorage temp_storage;
+
+  // Now we perform coalesced writes back to global memory using cub.
+  uint32_t* block_base = code_points + blockIdx.x * blockDim.x * MAX_NEW_CHARS;
+  BlockStore(temp_storage).Store(block_base, replacement_code_points);
+}
+
+// There are fixed lookup tables that are copied from codepoint_metadata.ah
```

Comment:
This that giant header file of integer values we talked about. I was able to reduce it from 19MB to less than 1MB but if clang-format gets its way, it will balloon up again. I used the `.ah` header so the style-check will ignore it. GH will not display large files like this one.

Reply:
Sneaky. :)

Reply:
That's pretty cool